### PR TITLE
[Feature] Add several MLU ops

### PR DESCRIPTION
**Pull request by teamwong111**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Refactor MLU csrc and Merge mlu to master.

## Modification

Modify setup.py and refactor MLU csrc and solve conflicts between master-branch and mlu-branch

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Review by grimoire — APPROVED** [review]
LGTM.

File: mmcv/ops/csrc/common/pytorch_mlu_helper.hpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
+/*************************************************************************
```
Comment (zhouzaida) [inline mmcv/ops/csrc/common/pytorch_mlu_helper.hpp:1]: Some definitions at this file can be found at `mmcv/ops/csrc/common/mlu/common_mlu_helper.hpp`.
Should we keep consistent with CUDA that `mmcv/ops/csrc/common/pytorch_cuda_helper.hpp` includes the `common_cuda_helper.hpp`. https://github.com/open-mmlab/mmcv/blob/48f44b697a3510803a252c78538a9064458995a5/mmcv/ops/csrc/common/pytorch_cuda_helper.hpp#L11

@grimoire

File: mmcv/utils/pytorch_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+import torch
+
+TORCH_VERSION = torch.__version__
+
+
+def is_cuda() -> bool:
+    return torch.cuda.is_available()
+
+
+def is_mlu() -> bool:
```
Comment (zhouzaida) [inline mmcv/utils/pytorch_wrapper.py:11]: Should we rename `is_mul` to `is_mlu_pytorch`

File: setup.py [hunk 3]
```diff
@@ -284,11 +288,27 @@ def get_extensions():
...
+                    include_dirs.append(
+                        os.path.abspath('./mmcv/ops/csrc/common'))
+                else:
+                    print('Cambricon Catch is not available!')
+            except AttributeError:
+                print(f'Compiling {ext_name} without CUDA')
+                op_files = glob.glob('./mmcv/ops/csrc/pytorch/*.cpp') + \
+                    glob.glob('./mmcv/ops/csrc/pytorch/cpu/*.cpp')
+                extension = CppExtension
+                include_dirs.append(os.path.abspath('./mmcv/ops/csrc/common'))
```
Comment (zhouzaida) [inline setup.py:311]: Maybe we can refactor the block as the following format
```python
#line 290
elif has_method(torch, 'is_mlu_available') and torch.is_mlu_available() or os.getenv('FORCE_MLU', '0') == '1':
    from torch_mlu.utils.cpp_extension import MLUExtension
    define_macros += [('MMCV_WITH_MLU', None)]
    mlu_args = os.getenv('MMCV_MLU_ARGS')
    extra_compile_args['cncc'] = [mlu_args] if mlu_args else []
    op_files = glob.glob('./mmcv/ops/csrc/pytorch/*.cpp')
    op_files += glob.glob('./mmcv/ops/csrc/pytorch/cpu/*.cpp')
    op_files += glob.glob('./mmcv/ops/csrc/pytorch/mlu/*.cpp')
    op_files += glob.glob('./mmcv/ops/csrc/common/mlu/*.mlu')
    extension = MLUExtension
    include_dirs.append(
        os.path.abspath('./mmcv/ops/csrc/common'))
else:
    print(f'Compiling {ext_name} only with CPU')
    op_files = glob.glob('./mmcv/ops/csrc/pytorch/*.cpp') + \
        glob.glob('./mmcv/ops/csrc/pytorch/cpu/*.cpp')
    extension = CppExtension
    include_dirs.append(os.path.abspath('./mmcv/ops/csrc/common'))

File: mmcv/ops/csrc/common/pytorch_mlu_helper.hpp (hunk 1 above)
Comment (grimoire) [reply mmcv/ops/csrc/common/pytorch_mlu_helper.hpp:1]: Yes, I think so.

File: mmcv/utils/pytorch_wrapper.py (hunk 2 above)
Comment (teamwong111) [reply mmcv/utils/pytorch_wrapper.py:11]: It is more accurate, but it seems to be inconsistent with `is_cuda`, so I don't think it should be renamed.

File: setup.py (hunk 3 above)
Comment (teamwong111) [reply setup.py]: Done

File: mmcv/ops/csrc/common/pytorch_mlu_helper.hpp (hunk 1 above)
Comment (teamwong111) [reply mmcv/ops/csrc/common/pytorch_mlu_helper.hpp:1]: Done

**Comment by zhouzaida** [comment]
#1770 had been merged. Please resolve conflicts.

File: mmcv/utils/pytorch_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+
+TORCH_VERSION = torch.__version__
```
Comment (zhouzaida) [inline mmcv/utils/pytorch_wrapper.py:4]: It is repeated with [here](https://github.com/open-mmlab/mmcv/blob/7982dd1a06bbcee3eb15a56aa55b8c18ebd2bf7c/mmcv/utils/parrots_wrapper.py#L6).

File: mmcv/utils/pytorch_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+
+
+def is_cuda() -> bool:
+    return torch.cuda.is_available()
```
Comment (zhouzaida) [inline mmcv/utils/pytorch_wrapper.py:8]: Add a variable `IS_CUDA = is_cuda()`.

File: tests/test_ops/test_nms.py [hunk 6]
```diff
@@ -3,12 +3,23 @@
...
 class Testnms(object):
 
-    def test_nms_allclose(self):
-        if not torch.cuda.is_available():
-            return
+    @pytest.mark.parametrize('device', [
+        pytest.param(
+            'cuda',
+            marks=pytest.mark.skipif(
+                not is_cuda(), reason='requires CUDA support')),
```
Comment (zhouzaida) [inline tests/test_ops/test_nms.py:16]:
```suggestion
                not IS_CUDA, reason='requires CUDA support')),
```

File: mmcv/ops/csrc/common/pytorch_cpp_helper.hpp [hunk 7]
```diff
@@ -11,7 +11,7 @@ using namespace at;
 #define CHECK_MLU(x) \
   TORCH_CHECK(x.device().type() == at::kMLU, #x " must be a MLU tensor")
 #define CHECK_CPU(x) \
-  TORCH_CHECK(!x.device().is_cuda(), #x " must be a CPU tensor")
+  TORCH_CHECK(x.device().type() == at::kCPU, #x " must be a CPU tensor")
```
Comment (Qiza-lyhm) [inline mmcv/ops/csrc/common/pytorch_cpp_helper.hpp:14]: Hi, it seems no mlu_bind.cpp files here. Please check your commit files?

File: setup.py [hunk 8]
```diff
@@ -302,12 +302,12 @@ def get_extensions():
             op_files = glob.glob('./mmcv/ops/csrc/pytorch/*.cpp') + \
                 glob.glob('./mmcv/ops/csrc/pytorch/cpu/*.cpp') + \
                 glob.glob('./mmcv/ops/csrc/pytorch/mlu/*.cpp') + \
-                glob.glob('./mmcv/ops/csrc/pytorch/mlu/*.mlu')
+                glob.glob('./mmcv/ops/csrc/common/mlu/*.mlu')
```
Comment (zhouzaida) [inline setup.py:305]: Should this line be removed?

Comment (zhouzaida) [inline setup.py:305]:
```suggestion
```

File: mmcv/ops/tin_shift.py [hunk 9]
```diff
@@ -18,6 +18,10 @@ class TINShiftFunction(Function):
 
     @staticmethod
     def forward(ctx, input, shift):
+        if input.size(0) != shift.size(0):
+            raise ValueError(
+                f'The first dim (batch) of `input` and `shift` should be'
```
Comment (ZwwWayne) [inline mmcv/ops/tin_shift.py:23]: unnecessary 'f' here

File: mmcv/runner/dist_utils.py [hunk 10]
```diff
@@ -49,7 +49,7 @@ def init_dist(launcher, backend='nccl', **kwargs):
 def _init_dist_pytorch(backend, **kwargs):
     # TODO: use local_rank instead of rank % num_gpus
     rank = int(os.environ['RANK'])
-    if IS_MLU:
+    if IS_MLU_AVAILABLE:
```
Comment (ZwwWayne) [inline mmcv/runner/dist_utils.py:52]: Seems that this name is inconsistent with IS_IPU

File: mmcv/utils/logging.py [hunk 11]
```diff
@@ -89,7 +89,7 @@ def print_log(msg, logger=None, level=logging.INFO):
         msg (str): The message to be logged.
         logger (logging.Logger | str | None): The logger to be used.
             Some special loggers are:
-            
+
```
Comment (ZwwWayne) [inline mmcv/utils/logging.py:92]: merge master and this has been fixed

File: mmcv/ops/tin_shift.py (hunk 9 above)
Comment (zhouzaida) [inline mmcv/ops/tin_shift.py:23]:
```suggestion
                'The first dim (batch) of `input` and `shift` should be '
```

File: setup.py [hunk 12]
```diff
@@ -302,12 +302,11 @@ def get_extensions():
             op_files = glob.glob('./mmcv/ops/csrc/pytorch/*.cpp') + \
                 glob.glob('./mmcv/ops/csrc/pytorch/cpu/*.cpp') + \
                 glob.glob('./mmcv/ops/csrc/pytorch/mlu/*.cpp') + \
```
Comment (zhouzaida) [inline setup.py:304]:
```suggestion
                glob.glob('./mmcv/ops/csrc/pytorch/mlu/*.cpp')
```

File: mmcv/runner/dist_utils.py (hunk 10 above)
Comment (zhouzaida) [reply mmcv/runner/dist_utils.py:52]: Their logic has some differences.

**Review by zhouzaida — APPROVED** [review]
